Shift watch

crosspolar opened the pull request:
Replacement for #612 .

What should be notified:

- “The shift is almost full.”
- “The shift is full.”
- ~~“A slot has opened up, or someone is looking for a replacement”~~ is not relevant enough!
- “The shift is now understaffed.”
-  "All clear: The shift is no longer understaffed, but it's not fully staffed yet either..."

### Missing:

- [x] Search for shifts which haven't been filled yet
- [x] [Optional] Select only certain events to be notified about.

### I don't like...
... the [for-looping all watched shifts](https://github.com/SuperCoopBerlin/tapir/pull/651/files#diff-611b38781ae92ec1fbed05624aaa95b4d12bda56e13e1c985832eaa752a16584R18-R27). I initially build a new model to store "Shift-staffing-Statuses", but it didn't feel good either.

Theophile-Madet reviewed (changes requested):
There's a few style points and the `send_shift_watch_mail` that is quite strange.

Did you talk to anyone who would use that function about how they would like to receive the notifications and about how much detail they want to have?

crosspolar commented on tapir/shifts/models.py:
```diff
@@ -1208,3 +1208,49 @@ class SolidarityShift(models.Model):
+    ALL_CLEAR = "AC", _(
+        "All clear: The shift is no longer understaffed, but it's not fully staffed yet either..."
+    )
+    __empty__ = _("(Unknown)")
```
I went by the [docs](https://docs.djangoproject.com/en/5.2/ref/models/fields/#enumeration-types):
> Because an enumeration with a concrete data type requires all values to match the type, overriding the [blank label](https://docs.djangoproject.com/en/5.2/ref/models/fields/#field-choices-blank-label) cannot be achieved by creating a member with a value of None. Instead, set the `__empty__` attribute on the class
> ```python
> class Answer(models.IntegerChoices):
>    NO = 0, _("No")
>    YES = 1, _("Yes")
>
>    __empty__ = _("(Unknown)")
> ```

crosspolar commented on tapir/shifts/utils.py:
```diff
@@ -164,3 +168,23 @@ def get_attendance_mode_display(attendance_mode: str) -> str:
+def get_staffing_status(shift: Shift, last_status: str):
+    if shift.get_valid_attendances().count() < shift.get_num_required_attendances():
+        return StaffingStatus.UNDERSTAFFED
+    elif shift.slots.count() - shift.get_valid_attendances().count() == 1:
+        return StaffingStatus.ALMOST_FULL
```
I see it from user's perspective and if one slot is free and I want to be in that shift, I want to know

crosspolar commented on tapir/shifts/management/commands/send_shift_watch_mail.py:
```diff
@@ -0,0 +1,44 @@
+
+    @staticmethod
+    def send_shift_watch_mail(shift_watches: ShiftWatch, reason: str):
+        for shift_watch in shift_watches.objects.select_related("user", "shift"):
```
It was from my old code in the other PR and I didn't notice the artifact, thanks

crosspolar commented on tapir/shifts/emails/shift_watch_mail.py:
```diff
@@ -0,0 +1,63 @@
+
+
+class ShiftWatchEmailBuilder(TapirEmailBuilderBase):
+    option = MailOption.OPTIONAL_ENABLED
```
It will be enabled by default for all users, but I like the option to disable it. For example, you follow some shifts and leave for vacation. Instead of un-following all you can just disable notifications during your absence

crosspolar commented:
> About for-looping all watched shifts, I don't think there is a way around it. Wouldn't that be like max 100 objects to loop through anyway? You can optimize the number of DB queries with a few `select_related` or `prefetch_related`, DB queries are usually what costs the most time. To test them I usually create a temporary view that just executes the function or command, and look at the analysis with Silk (`ENABLE_SILK_PROFILING` in the settings then go to `http://localhost:8000/silk/requests/`).
> 
> You could maybe do it all in one DB query with some smart annotations, filtering and subqueries, but it would be more complex, I think we benefit more from keeping it simple to understand.

I have now 
`for shift_watch_data in ShiftWatch.objects.select_related("user", "shift"):`, this sounds like it helps.

Theophile-Madet commented on tapir/shifts/tests/test_shiftwatch_forms.py:
```diff
@@ -0,0 +1,37 @@
+
+class TestShiftCancel(TapirFactoryTestBase):
+
+    @pytest.mark.django_db
```
I think we don't need this since we already inherit from `django.test.testcases.TestCase`

Theophile-Madet commented on tapir/shifts/tests/test_shiftwatch_forms.py:
```diff
@@ -0,0 +1,37 @@
+class TestShiftCancel(TapirFactoryTestBase):
+
+    @pytest.mark.django_db
+    def test_shift_watch_form_creates_shift_watch_instance(self):
```
Try to follow the [naming convention](https://github.com/SuperCoopBerlin/tapir/wiki/Automated-tests-%E2%80%90-overview#naming-convention)

Theophile-Madet commented on tapir/shifts/tests/test_shiftwatch_forms.py:
```diff
@@ -0,0 +1,37 @@
+            msg_prefix="The request should redirect to the shift's page.",
+        )
+
+        assert ShiftWatch.objects.count() == 1
```
Use `self.assert[...]` to stay consistent with the other tests

Theophile-Madet commented on tapir/shifts/views/views.py:
```diff
@@ -353,3 +357,31 @@ def get(self, request, *args, **kwargs):
         return super().get(request, args, kwargs)
+
+
+# class WatchShiftView(LoginRequiredMixin, RedirectView):
+#     def post(self, request, *args, **kwargs):
+#         shift = get_object_or_404(Shift, pk=kwargs["shift_id"])
+#         ShiftWatch.objects.get_or_create(user=request.user, shift=shift)
+#         return redirect("shifts:shift_detail", pk=kwargs["shift_id"])
+#
```
We can remove this

Theophile-Madet commented on tapir/shifts/views/views.py:
```diff
@@ -353,3 +357,31 @@ def get(self, request, *args, **kwargs):
+        return redirect("shifts:shift_detail", pk=kwargs["shift"])
+
+
+class WatchShiftView(LoginRequiredMixin, TapirFormMixin, CreateView):
```
The page title and the card title are missing, see an example in `tapir.coop.views.draftuser.DraftUserCreateView.get_context_data`

Theophile-Madet commented on tapir/shifts/models.py:
```diff
@@ -1208,3 +1208,66 @@ class SolidarityShift(models.Model):
+    )
+    shift = models.ForeignKey(Shift, on_delete=models.CASCADE)
+    staffing_events = ArrayField(
+        models.CharField(max_length=30, choices=get_staffingevent_choices),
```
Could this just be `choices=StaffingEventsChoices.choices`? I'm not sure what `get_staffingevent_choices` is doing.

Theophile-Madet commented on tapir/shifts/models.py:
```diff
@@ -1208,3 +1208,66 @@ class SolidarityShift(models.Model):
+        null=True,
+        choices=get_staffingevent_choices,
+    )
+    notification_timedelta = models.DurationField(default=datetime.timedelta(days=2))
```
This field is not used anywhere

crosspolar commented:
> I'm not sure the team leader notifications should be tied to ShiftWatch, since ShiftWatch is specifically to watch a single shift and the team leader notifications as I understand them would be for all shifts. For those I would rather add checks to RegisterUserToShiftSlotView and UpdateShiftAttendanceStateBase, what do you think?

I think it would give more flexibility to include it at the shift-watch. 
- Imagine for example one leading member is responsible for all shifts from Monday to Wednesday, and another one for all remaining days. In their personal accounts they could then.
- Per shift, we could also store in the model if leading-members want to watch only certain slots. 

I know, it sounds far ahead but maybe I couldn't find better examples :smile: 

In a future release, I would appreciate expand the feature to watch whole ShiftTemplates, so I can observe my own recurring shift.

crosspolar commented on tapir/shifts/management/commands/send_shift_watch_mail.py:
```diff
@@ -0,0 +1,106 @@
+                        shift_watch=shift_watch_data, reason=reason.label
+                    )
+
+            with transaction.atomic():
```
But isn't it two with `last_reason_for_notification` and  `shift_watch_data`?

crosspolar commented on tapir/shifts/management/commands/send_shift_watch_mail.py:
```diff
@@ -0,0 +1,106 @@
+                recipient=shift_watch.user,
+                email_builder=email_builder,
+            )
+            shift_watch.save()
```
you're right, I forgot deleting it

crosspolar commented:
My notes from call with @Theophile-Madet :

- no more features should be added to this PR to merge it sooner, just fixing your suggestions
- we will also deploy watching team leaders. Watching whole days or ShiftTemplates will follow later
- I'll have a look into FeatureFlag

crosspolar commented on tapir/shifts/templates/shifts/shift_block_tag.html:
```diff
@@ -10,6 +11,7 @@
             {% else %}
                 {{ shift.start_time|time:"H:i" }} - {{ shift.end_time|time:"H:i" }}
             {% endif %}
+            {% if request.user|user_watching_shift:shift.id %}<div class="material-icons">visibility</div>{% endif %}
```
I tried using annotate, what do you think? https://github.com/SuperCoopBerlin/tapir/pull/651/commits/85cc38e0a3905182c11955a3c092f9e8464827c2

crosspolar commented:
I've reformatted the `send_shift_watch_mail` and adapted your suggestions (most of it). Also please have a look at the `annotate` I added. I tried seeing changes with silk, but maybe test-data was too small seeing a difference.
I've added one small feature: Initial selection of possible Choices is what has been chosen the last time, if available. So that if you want to watch a bunch of shifts, you don't have to select everytime the same.

crosspolar commented on tapir/shifts/management/commands/send_understaffed_shift_reminder_mail.py:
```diff
@@ -0,0 +1,51 @@
+        shifts_tomorrow = ShiftWatch.objects.filter(shift__end_time__date=tomorrow)
+
+        for shift_watch_data in shifts_tomorrow:
+            this_valid_slot_ids = [
+                s.slot_id for s in shift_watch_data.shift.get_valid_attendances()
+            ]
+            number_of_available_slots = shift_watch_data.shift.slots.count()
+            valid_attendances_count = len(this_valid_slot_ids)
+            required_attendances_count = (
+                shift_watch_data.shift.get_num_required_attendances()
+            )
+            current_status = get_staffing_status(
+                number_of_available_slots=number_of_available_slots,
+                valid_attendances=valid_attendances_count,
+                required_attendances=required_attendances_count,
+                last_status=shift_watch_data.last_staffing_status,
+            )
```
I mean, currently `send_shift_watch_mail_per_user_and_shift` also checks for teamleader and other status and saves a status them. Here, we only want to go after Understaffed
